From c833085e2a9280aa1b049cdb2452795599279991 Mon Sep 17 00:00:00 2001
From: Steven Koo <steven.koo@ni.com>
Date: Fri, 1 Apr 2022 10:40:14 -0500
Subject: [PATCH 17/18] sulfur: power: fix reboot on powerdown

Often on powerdown, the device reboots because the watchdog
is triggered during the shutdown sequence.

This commit introduces a new flag wdt_enabled
which decides if a watchdog event triggers a
warm reset or is ignored.

This is based on the neon version of the patch

Co-authored-by: Joerg Hofrichter <joerg.hofrichter@ni.com>
Signed-off-by: Steven Koo <steven.koo@ni.com>
---
 board/sulfur/power.c | 16 ++++++++++++----
 1 file changed, 12 insertions(+), 4 deletions(-)

diff --git a/board/sulfur/power.c b/board/sulfur/power.c
index d0cb64612b..2c4e6078d1 100644
--- a/board/sulfur/power.c
+++ b/board/sulfur/power.c
@@ -53,6 +53,7 @@
 
 static int forcing_shutdown;
 static int wdt_reset;
+static int wdt_enabled;
 static int is_rev5;
 static int is_rev10;
 
@@ -175,6 +176,7 @@ enum power_state power_handle_state(enum power_state state)
 {
 	switch (state) {
 	case POWER_G3:
+		wdt_enabled = 0;
 		break;
 	case POWER_S5:
 		if (forcing_shutdown)
@@ -194,6 +196,7 @@ enum power_state power_handle_state(enum power_state state)
 		gpio_set_level(GPIO_POWER_EN_IO, 1);
 		usleep(5);
 		gpio_set_level(GPIO_POWER_EN_3V7, 1);
+		wdt_enabled = 1;
 		usleep(5);
 
 		if (power_wait_signals_timeout(IN_PGOOD_S3, 100 * MSEC)
@@ -302,6 +305,7 @@ enum power_state power_handle_state(enum power_state state)
 		gpio_set_level(GPIO_POWER_EN_1V5, 0);
 		gpio_set_level(GPIO_POWER_EN_IO, 0);
 		gpio_set_level(GPIO_POWER_EN_3V7, 0);
+		wdt_enabled = 0;
 
 		/*
 		 * Enable idle task deep sleep. Allow the low power idle task
@@ -389,8 +393,12 @@ DECLARE_HOOK(HOOK_TICK, reset_button_poll, HOOK_PRIO_DEFAULT);
 
 void wdt_reset_event(enum gpio_signal signal)
 {
-	CPRINTS("Watchdog timeout, warm reset the AP");
-	wdt_reset = 1;
-	host_set_single_event(EC_HOST_EVENT_HANG_REBOOT);
-	hook_call_deferred(&force_reset_data, 10 * MSEC);
+	if (wdt_enabled) {
+		CPRINTS("Watchdog timeout, warm reset the AP");
+		wdt_reset = 1;
+		host_set_single_event(EC_HOST_EVENT_HANG_REBOOT);
+		hook_call_deferred(&force_reset_data, 10 * MSEC);
+	} else {
+		CPRINTS("Watchdog timeout, ignored");
+	}
 }
-- 
2.43.0

